csrc/scheduler/normalization_inner_tma.cpp

```diff
@@ -5,62 +5,296 @@
+    reduction_tv->axis(ipos + 1)->parallelize(ParallelType::TIDy);
+    rpos = ipos + 2;
+  } else {
+    reduction_tv->axis(ipos)->parallelize(ParallelType::BIDx);
+  }
+  reduction_tv->axis(rpos)->parallelize(ParallelType::Bulk);
+  scheduler_utils::parallelizeAllLike(reduction_tv, tma_tvs);
+  // Reset reduction_tv's reduction axis back to Serial (only TMA loads use
+  // Bulk)
+  reduction_tv->axis(rpos)->parallelize(ParallelType::Serial);
+
+  // Transform reduction domain for efficient computation:
+  //   [I, R] -> [I, b, us, x, v]
+  // Where:
+  //   - v: vectorization factor (elements per vector instruction)
+  //   - x: thread dimension (TIDx, threads cooperating on reduction)
+  //   - b: persistent batch size (register persistent buffer size)
+  //   - us: unswitch dimension (loop optimization, reduces control flow
+  //   overhead)
+  reduction_tv->split(rpos, params->vectorization_factor);
+  reduction_tv->split(rpos, params->persistent_batch_size, false);
+  reduction_tv->split(rpos, 1);
+  reduction_tv->axis(rpos + 1)->parallelize(ParallelType::Unswitch);
+  reduction_tv->axis(rpos + 2)->parallelize(ParallelType::TIDx);
+  reduction_tv->axis(rpos + 2)->padToMultipleOfWarp();
+
+  // Create rfactor tensor to separate thread-local reduction from block
+  // reduction This enables a two-stage reduction:
+  //   1. Thread-local vectorized reduction (across b and v dimensions)
+  //   2. Block-level reduction (across x dimension using warp/block primitives)
+  // rfactor axes: {rpos, vectorize_pos} corresponding to b and v dimensions
+  int64_t vectorize_pos = rpos + 3;
+  auto reference_tv = reduction_tv->rFactor({rpos, vectorize_pos});
+
+  // Propagate transformations from reference_tv to all non-TMA tensors
+  // TMA tensors keep their simple [BIDx, Bulk] schedule
+  std::vector<TensorView*> non_tma_tvs =
+      ir_utils::allTvsExcept(fusion, {tma_tvs.begin(), tma_tvs.end()});
+  TransformPropagator non_tma_propagator(reference_tv);
+  SetSelector selector({non_tma_tvs.begin(), non_tma_tvs.end()});
+  MaxLogicalDomainInfoSpanningTree(reference_tv, &selector)
+      .traverse(&non_tma_propagator);
+
+  // If reduction_tv is rfactored, rfactor all reductions.
+  // Also needs to update non_tma_tvs to include newly rfactored tvs.
+  if (reference_tv != reduction_tv) {
+    reduction_scheduler_utils::propagateRFactor(
+        reference_tv, reduction_tv, reduction_tvs);
+    non_tma_tvs =
+        ir_utils::allTvsExcept(fusion, {tma_tvs.begin(), tma_tvs.end()});
+  }
+  scheduler_utils::parallelizeAllLike(reference_tv, non_tma_tvs);
+
+  // Helper lambda to find the vectorization position (one past TIDx axis)
+  auto get_vect_pos = [](TensorView* tv) {
+    auto it = std::find_if(
+        tv->domain()->loop().begin(),
+        tv->domain()->loop().end(),
+        [](const IterDomain* id) {
+          return id->getParallelType() == ParallelType::TIDx;
+        });
+
+    if (it == tv->domain()->loop().end()) {
+      return -1;
```

Comment:
`-1` is a valid axis position. 

Why not return `std::nullopt` and check `output.has_value()`? You'd need to update lambda signature to `auto get_vect_pos = [](TensorView* tv) -> std::optional<int64_t> {` if you return `std::optional`.